File: notes/ios/Sources/Notes/APIClient.swift
import Foundation

enum APIError: Error, LocalizedError {
    case unauthorized
    case notFound
    case conflict(String)
    case server(Int)

    var errorDescription: String? {
        switch self {
        case .unauthorized:      return "Invalid email or password"
        case .notFound:          return "Not found"
        case .conflict(let msg): return msg
        case .server(let c):     return "Server error (\(c))"
        }
    }
}

final class APIClient {
    var serverBase: String
    var token: String?

    private let decoder = JSONDecoder()
    private let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest  = 10
        cfg.timeoutIntervalForResource = 30
        return URLSession(configuration: cfg)
    }()

    init(serverBase: String, token: String? = nil) {
        self.serverBase = serverBase
        self.token      = token
    }

    // MARK: - Auth (unauthenticated)

    func register(email: String, password: String) async throws -> AuthResponse {
        try await post("/auth/register",
                       body: ["email": email, "password": password],
                       authenticated: false)
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        try await post("/auth/login",
                       body: ["email": email, "password": password],
                       authenticated: false)
    }

    func recover(email: String, code: String, newPassword: String) async throws -> AuthResponse {
        try await post("/auth/recover",
                       body: ["email": email, "recovery_code": code, "new_password": newPassword],
                       authenticated: false)
    }

    // MARK: - Folders

    func listFolders() async throws -> [Folder] {
        struct R: Decodable { let folders: [Folder] }
        let r: R = try await get("/folders")
        return r.folders
    }

    func createFolder(id: String, name: String, createdAt: Int64) async throws -> Folder {
        struct R: Decodable { let folder: Folder }
        let r: R = try await post("/folders",
                                  body: ["id": id, "name": name, "created_at": createdAt])
        return r.folder
    }

    func renameFolder(id: String, name: String) async throws -> Folder {
        struct R: Decodable { let folder: Folder }
        let r: R = try await patch("/folders/\(id)", body: ["name": name])
        return r.folder
    }

    func deleteFolder(id: String) async throws {
        try await delete("/folders/\(id)")
    }

    // MARK: - Notes

    func listNotes() async throws -> [Note] {
        struct R: Decodable { let notes: [Note] }
        let r: R = try await get("/notes")
        return r.notes
    }

    func createNote(
        id: String, folderID: String?,
        title: String, content: String,
        updatedAt: Int64, createdAt: Int64
    ) async throws -> Note {
        var body: [String: Any] = [
            "id": id, "title": title, "content": content,
            "updated_at": updatedAt, "created_at": createdAt,
        ]
        if let fid = folderID, !fid.isEmpty { body["folder_id"] = fid }
        struct R: Decodable { let note: Note }
        let r: R = try await post("/notes", body: body)
        return r.note
    }

    func updateNote(id: String, title: String, content: String, updatedAt: Int64) async throws -> Note {
        struct R: Decodable { let note: Note }
        let r: R = try await patch("/notes/\(id)",
                                   body: ["title": title, "content": content, "updated_at": updatedAt])
        return r.note
    }

    func moveNote(id: String, folderID: String?) async throws -> Note {
        struct R: Decodable { let note: Note }
        let r: R = try await patch("/notes/\(id)/move", body: ["folder_id": folderID ?? ""])
        return r.note
    }

    func trashNote(id: String) async throws {
        try await performVoid(try makeRequest(method: "PATCH", path: "/notes/\(id)/trash"))
    }

    func restoreNote(id: String) async throws -> Note {
        struct R: Decodable { let note: Note }
        let r: R = try await perform(try makeRequest(method: "PATCH", path: "/notes/\(id)/restore"))
        return r.note
    }

    func deleteNote(id: String) async throws {
        try await delete("/notes/\(id)")
    }

    // MARK: - WebSocket URL

    func wsURL(forNoteID noteID: String) -> URL? {
        guard let tok = token else { return nil }
        let wsBase = serverBase
            .replacingOccurrences(of: "https://", with: "wss://")
            .replacingOccurrences(of: "http://",  with: "ws://")
        return URL(string: "\(wsBase)/notes/\(noteID)/ws?token=\(tok)")
    }

    // MARK: - Private helpers

    private func makeRequest(
        method: String,
        path: String,
        body: [String: Any]? = nil,
        authenticated: Bool = true
    ) throws -> URLRequest {
        guard let url = URL(string: serverBase + path) else { throw URLError(.badURL) }
        var req = URLRequest(url: url)
        req.httpMethod = method
        if authenticated, let tok = token {
            req.setValue("Bearer \(tok)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        return req
    }

    private func checkStatus(_ response: URLResponse, body: Data) throws {
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        switch http.statusCode {
        case 200, 201, 204: return
        case 401: throw APIError.unauthorized
        case 404: throw APIError.notFound
        case 409: throw APIError.conflict(serverMessage(body) ?? "Conflict")
        default:  throw APIError.server(http.statusCode)
        }
    }

    private func serverMessage(_ data: Data) -> String? {
        let s = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
        return (s?.isEmpty == false) ? s : nil
    }

    private func perform<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: req)
        try checkStatus(response, body: data)
        return try decoder.decode(T.self, from: data)
    }

    private func performVoid(_ req: URLRequest) async throws {
        let (data, response) = try await session.data(for: req)
        try checkStatus(response, body: data)
    }

    private func get<T: Decodable>(_ path: String) async throws -> T {
        try await perform(try makeRequest(method: "GET", path: path))
    }

    private func post<T: Decodable>(
        _ path: String,
        body: [String: Any],
        authenticated: Bool = true
    ) async throws -> T {
        try await perform(try makeRequest(method: "POST", path: path, body: body, authenticated: authenticated))
    }

    private func patch<T: Decodable>(_ path: String, body: [String: Any]) async throws -> T {
        try await perform(try makeRequest(method: "PATCH", path: path, body: body))
    }

    private func delete(_ path: String) async throws {
        try await performVoid(try makeRequest(method: "DELETE", path: path))
    }
}

File: notes/linux/src/backend/api_client.rs
use anyhow::Result;
use reqwest::{Client, StatusCode};
use serde::Deserialize;

use crate::backend::local_store::{FolderRecord, NoteRecord};

// ── Error type ────────────────────────────────────────────────────────────────

#[derive(Debug, thiserror::Error)]
pub enum ApiError {
    #[error("Invalid email or password")]
    Unauthorized,
    #[error("Not found")]
    NotFound,
    #[error("Conflict: {0}")]
    Conflict(String),
    #[error("Server error ({0})")]
    Server(u16),
    #[error(transparent)]
    Http(#[from] reqwest::Error),
}

// ── Wire types ────────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct AuthResponse {
    pub token: String,
    pub recovery_code: Option<String>,
}

#[derive(Debug, Deserialize)]
struct FoldersResponse {
    folders: Vec<FolderRecord>,
}

#[derive(Debug, Deserialize)]
struct FolderResponse {
    folder: FolderRecord,
}

#[derive(Debug, Deserialize)]
struct NotesResponse {
    notes: Vec<NoteRecord>,
}

#[derive(Debug, Deserialize)]
struct NoteResponse {
    note: NoteRecord,
}

// ── Client ────────────────────────────────────────────────────────────────────

#[derive(Clone)]
pub struct ApiClient {
    http: Client,
    pub base_url: String,
    pub token: Option<String>,
}

impl ApiClient {
    pub fn new(base_url: String, token: Option<String>) -> Self {
        Self {
            http: Client::builder()
                .timeout(std::time::Duration::from_secs(30))
                .build()
                .expect("failed to build reqwest client"),
            base_url,
            token,
        }
    }

    pub fn ws_url_for_note(&self, note_id: &str) -> Option<String> {
        let tok = self.token.as_ref()?;
        let ws_base = self
            .base_url
            .replace("https://", "wss://")
            .replace("http://", "ws://");
        Some(format!("{ws_base}/notes/{note_id}/ws?token={tok}"))
    }

    // ── Auth ──────────────────────────────────────────────────────────────────

    pub async fn login(&self, email: &str, password: &str) -> Result<AuthResponse, ApiError> {
        self.post_unauth(
            "/auth/login",
            serde_json::json!({ "email": email, "password": password }),
        )
        .await
    }

    pub async fn register(&self, email: &str, password: &str) -> Result<AuthResponse, ApiError> {
        self.post_unauth(
            "/auth/register",
            serde_json::json!({ "email": email, "password": password }),
        )
        .await
    }

    pub async fn recover(
        &self,
        email: &str,
        code: &str,
        new_password: &str,
    ) -> Result<AuthResponse, ApiError> {
        self.post_unauth(
            "/auth/recover",
            serde_json::json!({
                "email": email,
                "recovery_code": code,
                "new_password": new_password,
            }),
        )
        .await
    }

    // ── Folders ───────────────────────────────────────────────────────────────

    pub async fn list_folders(&self) -> Result<Vec<FolderRecord>, ApiError> {
        let r: FoldersResponse = self.get("/folders").await?;
        Ok(r.folders)
    }

    pub async fn create_folder(
        &self,
        id: &str,
        name: &str,
        created_at: i64,
    ) -> Result<FolderRecord, ApiError> {
        let r: FolderResponse = self
            .post(
                "/folders",
                serde_json::json!({ "id": id, "name": name, "created_at": created_at }),
            )
            .await?;
        Ok(r.folder)
    }

    pub async fn rename_folder(&self, id: &str, name: &str) -> Result<FolderRecord, ApiError> {
        let r: FolderResponse = self
            .patch(&format!("/folders/{id}"), serde_json::json!({ "name": name }))
            .await?;
        Ok(r.folder)
    }

    pub async fn delete_folder(&self, id: &str) -> Result<(), ApiError> {
        self.delete(&format!("/folders/{id}")).await
    }

    // ── Notes ─────────────────────────────────────────────────────────────────

    pub async fn list_notes(&self) -> Result<Vec<NoteRecord>, ApiError> {
        let r: NotesResponse = self.get("/notes").await?;
        Ok(r.notes)
    }

    pub async fn create_note(&self, note: &NoteRecord) -> Result<NoteRecord, ApiError> {
        let mut body = serde_json::json!({
            "id": note.id,
            "title": note.title,
            "content": note.content,
            "updated_at": note.updated_at,
            "created_at": note.created_at,
        });
        if let Some(fid) = &note.folder_id {
            if !fid.is_empty() {
                body["folder_id"] = serde_json::Value::String(fid.clone());
            }
        }
        let r: NoteResponse = self.post("/notes", body).await?;
        Ok(r.note)
    }

    pub async fn update_note(
        &self,
        id: &str,
        title: &str,
        content: &str,
        updated_at: i64,
    ) -> Result<NoteRecord, ApiError> {
        let r: NoteResponse = self
            .patch(
                &format!("/notes/{id}"),
                serde_json::json!({
                    "title": title,
                    "content": content,
                    "updated_at": updated_at,
                }),
            )
            .await?;
        Ok(r.note)
    }

    pub async fn move_note(&self, id: &str, folder_id: Option<&str>) -> Result<NoteRecord, ApiError> {
        let r: NoteResponse = self
            .patch(
                &format!("/notes/{id}/move"),
                serde_json::json!({ "folder_id": folder_id.unwrap_or("") }),
            )
            .await?;
        Ok(r.note)
    }

    pub async fn trash_note(&self, id: &str) -> Result<(), ApiError> {
        self.patch_empty(&format!("/notes/{id}/trash")).await
    }

    pub async fn restore_note(&self, id: &str) -> Result<(), ApiError> {
        self.patch_empty(&format!("/notes/{id}/restore")).await
    }

    pub async fn delete_note(&self, id: &str) -> Result<(), ApiError> {
        self.delete(&format!("/notes/{id}")).await
    }

    // ── Blobs ─────────────────────────────────────────────────────────────────

    pub async fn upload_blob(&self, id: &str, data: Vec<u8>) -> Result<(), ApiError> {
        let url = format!("{}/blobs/{}", self.base_url, id);
        let mut req = self.http.put(&url).body(data);
        if let Some(auth) = self.auth_header() {
            req = req.header("Authorization", auth);
        }
        let resp = req.send().await?;
        match resp.status() {
            s if s.is_success() => Ok(()),
            StatusCode::UNAUTHORIZED => Err(ApiError::Unauthorized),
            s => Err(ApiError::Server(s.as_u16())),
        }
    }

    pub async fn download_blob(&self, id: &str) -> Result<Vec<u8>, ApiError> {
        let url = format!("{}/blobs/{}", self.base_url, id);
        let resp = self.http.get(&url).send().await?;
        match resp.status() {
            s if s.is_success() => Ok(resp.bytes().await?.to_vec()),
            StatusCode::NOT_FOUND => Err(ApiError::NotFound),
            s => Err(ApiError::Server(s.as_u16())),
        }
    }

    // ── HTTP helpers ──────────────────────────────────────────────────────────

    fn auth_header(&self) -> Option<String> {
        self.token.as_ref().map(|t| format!("Bearer {t}"))
    }

    async fn get<T: for<'de> Deserialize<'de>>(&self, path: &str) -> Result<T, ApiError> {
        let mut req = self.http.get(format!("{}{}", self.base_url, path));
        if let Some(auth) = self.auth_header() {
            req = req.header("Authorization", auth);
        }
        let resp = req.send().await?;
        self.parse(resp).await
    }

    async fn post<T: for<'de> Deserialize<'de>>(
        &self,
        path: &str,
        body: serde_json::Value,
    ) -> Result<T, ApiError> {
        let mut req = self
            .http
            .post(format!("{}{}", self.base_url, path))
            .json(&body);
        if let Some(auth) = self.auth_header() {
            req = req.header("Authorization", auth);
        }
        let resp = req.send().await?;
        self.parse(resp).await
    }

    async fn post_unauth<T: for<'de> Deserialize<'de>>(
        &self,
        path: &str,
        body: serde_json::Value,
    ) -> Result<T, ApiError> {
        let resp = self
            .http
            .post(format!("{}{}", self.base_url, path))
            .json(&body)
            .send()
            .await?;
        self.parse(resp).await
    }

    async fn patch<T: for<'de> Deserialize<'de>>(
        &self,
        path: &str,
        body: serde_json::Value,
    ) -> Result<T, ApiError> {
        let mut req = self
            .http
            .patch(format!("{}{}", self.base_url, path))
            .json(&body);
        if let Some(auth) = self.auth_header() {
            req = req.header("Authorization", auth);
        }
        let resp = req.send().await?;
        self.parse(resp).await
    }

    /// PATCH that expects 204 No Content (no response body).
    async fn patch_empty(&self, path: &str) -> Result<(), ApiError> {
        let mut req = self
            .http
            .patch(format!("{}{}", self.base_url, path))
            .header("Content-Length", "0");
        if let Some(auth) = self.auth_header() {
            req = req.header("Authorization", auth);
        }
        let resp = req.send().await?;
        match resp.status() {
            s if s.is_success() => Ok(()),
            StatusCode::UNAUTHORIZED => Err(ApiError::Unauthorized),
            StatusCode::NOT_FOUND => Err(ApiError::NotFound),
            s => Err(ApiError::Server(s.as_u16())),
        }
    }

    async fn delete(&self, path: &str) -> Result<(), ApiError> {
        let mut req = self.http.delete(format!("{}{}", self.base_url, path));
        if let Some(auth) = self.auth_header() {
            req = req.header("Authorization", auth);
        }
        let resp = req.send().await?;
        match resp.status() {
            StatusCode::NO_CONTENT | StatusCode::OK => Ok(()),
            StatusCode::UNAUTHORIZED => Err(ApiError::Unauthorized),
            StatusCode::NOT_FOUND => Err(ApiError::NotFound),
            s => Err(ApiError::Server(s.as_u16())),
        }
    }

    async fn parse<T: for<'de> Deserialize<'de>>(
        &self,
        resp: reqwest::Response,
    ) -> Result<T, ApiError> {
        match resp.status() {
            s if s.is_success() => Ok(resp.json::<T>().await?),
            StatusCode::UNAUTHORIZED => Err(ApiError::Unauthorized),
            StatusCode::NOT_FOUND => Err(ApiError::NotFound),
            StatusCode::CONFLICT => {
                let msg = resp.text().await.unwrap_or_default();
                Err(ApiError::Conflict(msg))
            }
            s => Err(ApiError::Server(s.as_u16())),
        }
    }
}
